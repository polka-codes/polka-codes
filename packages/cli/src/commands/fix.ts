// generated by polka.codes
import { Command } from 'commander'
import { runWorkflow } from '../runWorkflow'
import { type FixWorkflowInput, fixWorkflow } from '../workflows/fix.workflow'

export const fixCommand = new Command('fix')
  .description('Fix issues by running a command and letting an agent fix it.')
  .argument('[command]', 'The command to run')
  .option('-t, --task <task>', 'The task to perform, providing specific instructions to the agent.')
  .action(async (command: string | undefined, options: { task?: string }, cmd: Command) => {
    const workflowInput: FixWorkflowInput = { command, task: options.task }
    await runWorkflow('fix', fixWorkflow, cmd, workflowInput)
  })
