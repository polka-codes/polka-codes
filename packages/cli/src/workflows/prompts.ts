// generated by polka.codes

export function getWorkflowContextPrompt(): string {
  return `# Code Workflow Context

You are participating in a 3-phase development workflow:

1. **Planning Phase** (Architect Agent)
   - Analyzes requirements and creates detailed implementation plan
   - User reviews and approves the plan before implementation begins

2. **Implementation Phase** (Coder Agent)
   - Implements the approved plan from phase 1
   - Makes code changes following the plan and project conventions

3. **Fixing Phase** (CodeFixer Agent)
   - Runs validation commands (tests, type checks, linting)
   - Fixes any errors that arise during validation
   - Ensures code quality and correctness

`
}

export const EPIC_TASK_BREAKDOWN_PROMPT = `
You are an expert at breaking down a large project into smaller, manageable tasks.
Based on the provided high-level plan, create a sequence of smaller, implementable tasks and a brief technical overview of the epic.
The overview should be a short paragraph that summarizes the overall technical approach.
Each task should be a self-contained unit of work that can be implemented and committed separately.
Respond with a JSON object containing an "overview" string and a "tasks" array. Each object in the array should have a "description".

Example format:
\`\`\`json
{
  "overview": "This epic will introduce a new caching layer to improve performance. It involves setting up a Redis client, implementing a cache-aside pattern for data fetching, and adding configuration for the cache.",
  "tasks": [
    {
      "description": "Set up the initial project structure."
    },
    {
      "description": "Implement the main feature."
    }
  ],
  "branchName": "feat/new-caching-layer"
}
\`\`\`
`

export const EPIC_WORKFLOW_ORCHESTRATOR_PROMPT = `
You are the orchestrator of an epic workflow.
Your job is to guide the user and the AI agents through the entire process of implementing a large feature.
You will be responsible for:
1.  Gathering and clarifying requirements.
2.  Generating a high-level plan.
3.  Breaking down the plan into smaller tasks.
4.  Orchestrating the execution of each task.
5.  Ensuring the successful completion of the epic.
`

export const COMMIT_MESSAGE_PROMPT = `
You are an expert at writing git commit messages.
Based on the provided list of staged files in <file_status>, the diff in <diff> and optional user context in <tool_input_context>, generate a concise and descriptive commit message.

Follow the conventional commit format.

Respond with a JSON object containing the commit message.
Example format:
\`\`\`json
{
  "commitMessage": "feat: add new feature\\n\\ndescribe the new feature in more detail"
}
\`\`\`
`

export const INIT_WORKFLOW_ANALYZE_PROMPT = `
Role: Analyzer agent
Goal: Produce a valid polkacodes YAML configuration for the project.

Workflow
1. Scan project files to identify the project's characteristics. Start using the "readFile" tool to understand the project's dependencies, scripts, and basic configuration.
   - Package/build tool (npm, bun, pnpm, etc.)
   - Test framework and patterns (snapshot tests, coverage, etc.)
   - Formatter / linter and their rules
   - Folder structure and naming conventions.
   - CI / development workflows (e.g., GitHub Actions in .github/workflows).

2. Build a YAML config with three root keys:

\`\`\`yaml
scripts:          # derive from package.json and CI workflows. Only include scripts that are relevant for development.
  format:        # code formatter
    command: "<formatter cmd>"
    description: "Format code"
  check:         # linter / type checker
    command: "<linter cmd>"
    description: "Static checks"
  test:          # test runner
    command: "<test cmd>"
    description: "Run tests"
  # add any other meaningful project scripts like 'build', 'dev', etc.

rules:            # A bullet list of key conventions, frameworks, and libraries used (e.g., "- React", "- TypeScript", "- Jest"). This helps other agents understand the project.

excludeFiles:     # A list of glob patterns for files that should not be read. Only include files that might contain secrets.
  - ".env"
  - ".env.*"
  - "*.pem"
  - "*.key"
  - ".npmrc"
  # do NOT list build artifacts, lockfiles, or paths already in .gitignore
\`\`\`

3. Return a JSON object with the generated YAML configuration as a string in the 'yaml' property.

\`\`\`json
{
  "yaml": "<yaml_string>"
}
\`\`\`
`

export const META_PROMPT = `
You are a meta-agent that decides which workflow to use for a given task.
Based on the user's task, decide whether to use the 'code', or 'task' workflow.

- Use the 'code' workflow for tasks that are well-defined and can be implemented directly without a separate planning phase.
- Use the 'task' workflow for simple, single-action tasks like answering a question or running a command.

The user's task is provided in the <task> tag.

Respond with a JSON object in a markdown block with the following structure:
{
  "workflow": "<workflow_name>" // 'code' or 'task'
}
`

export const PLAN_PROMPT = `
# Role and Objective

You are an expert planner. Your goal is to create a detailed plan for a given task.

# Task Input

The user has provided a task:
<task>
{task}
</task>

{planContent}

# Plan Format Guidelines

When generating your plan, follow these formatting guidelines:

1. Number major sections to provide clear structure:
   a. Use numbers (1., 2., 3., etc.) for top-level sections
   b. Use nested numbering (1.1, 1.2) or letters (a., b., c.) for sub-sections
   c. This makes sections easy to reference and understand
   d. Provides clear hierarchy and organization

   Example section numbering:
   1. Project Setup
      1.1 Initialize repository
      1.2 Configure dependencies
   2. Implementation
      2.1 Core features
      2.2 Tests

2. Use numbered lists when the order of steps matters:
   a. Sequential steps where one depends on the previous
   b. Steps that must be performed in a specific order
   c. Processes with clear progression
   d. When steps need to be referenced by number

   Example numbered list format:
   1. First step that must be completed first
   2. Second step that depends on the first
   3. Third step that follows from the second

3. Use checklist format (markdown checkboxes) when the plan involves:
   a. Multiple independent action items
   b. Trackable items that can be marked as complete
   c. Verifiable completion criteria
   d. Tasks that benefit from progress tracking

   Example checklist format:
   - [ ] First action item
   - [ ] Second action item
   - [ ] Third action item

4. Use narrative or structured text format when the plan involves:
   a. High-level strategies or conceptual approaches
   b. Explanations or background information
   c. Decision-making guidance
   d. Context that doesn't translate well to discrete steps

5. Combine formats when appropriate:
   a. Use numbered sections for overall structure
   b. Use narrative text for context and explanation
   c. Use numbered lists for sequential steps
   d. Use checklist items for trackable actions

   Example combined format:
   1. Phase 1: Setup
      First, we need to configure the environment...
      1. Install dependencies
      2. Configure settings
      3. Verify installation

   2. Phase 2: Implementation
      - [ ] Implement feature A
      - [ ] Implement feature B
      - [ ] Write tests

6. Only include relevant details for AI Agents:
    a. Avoid unnecessary technical jargon or implementation details
    b. Avoid steps requires human intervention or cannot be done by an AI agent

# Your Tasks

1. Analyze the task and the existing plan (if any).
2. If the requirements are clear and you can generate or update the plan:
   a. Provide the plan in the "plan" field
   b. Apply appropriate formatting based on guidelines
3. If the requirements are not clear:
   a. Ask a clarifying question in the "question" field
4. If the task is already implemented or no action is needed:
   a. Do not generate a plan.
   b. Provide a concise reason in the "reason" field.

# Response Format

Respond with a JSON object that matches the following schema:
\`\`\`json
{
  "plan": "The generated or updated plan.",
  "question": "The clarifying question to ask the user.",
  "reason": "If no plan is needed, provide a reason here.",
  "files": ["path/to/file1.ts", "path/to/file2.ts"]
}
\`\`\`
`

export const GET_PR_DETAILS_PROMPT = `
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
`

export const CODE_REVIEW_PROMPT = `
# Code Review Prompt

You are a senior software engineer reviewing code changes.

## Critical Instructions
- **ONLY review the actual changes shown in the diff.** Do not comment on existing code that wasn't modified.
- **ONLY run gitDiff on files that are reviewable source/config files** per the "File Selection for gitDiff" rules below. Do not pass excluded files to gitDiff.

## File Selection for gitDiff
Use <file_status> to decide which files to diff. Include only files likely to contain human-authored source or meaningful configuration.

Include (run gitDiff):
- Application/source code
- UI/templates/assets code
- Infra/config that affects behavior

Exclude (do NOT run gitDiff; do not review):
- Lockfiles
- Generated/build artifacts & deps
- Test artifacts/snapshots
- Data and fixtures
- Binary/media/minified/maps

## Viewing Changes
- For each included file, **use gitDiff** to inspect the actual code changes:
  - **Pull request:** use the provided commit range for the gitDiff tool with contextLines: 5 and includeLineNumbers: true, but only surface and review the included files.
  - **Local changes:** diff staged or unstaged included files using gitDiff with contextLines: 5 and includeLineNumbers: true.
- The diff will include line number annotations: [Line N] for additions and [Line N removed] for deletions.
- You may receive:
  - <pr_title>
  - <pr_description>
  - <commit_messages>
- A <review_instructions> tag tells you the focus of the review.
- Use <file_status> to understand which files were modified, added, deleted, or renamed and to apply the inclusion/exclusion rules above.

## Line Number Reporting
- Use the line numbers from the annotations in the diff output.
- For additions: use the number from the [Line N] annotation after the + line.
- For deletions: use the number from the [Line N removed] annotation after the - line.
- For modifications: report the line number of the new/current code (from [Line N]).
- Report single lines as "N" and ranges as "N-M".

## Review Guidelines
Focus exclusively on the changed lines (+ additions, - deletions, modified lines):
- **Specific issues:** Point to exact problems in the changed code with accurate line references from the annotations.
- **Actionable fixes:** Provide concrete solutions, not vague suggestions.
- **Clear reasoning:** Explain why each issue matters and how to fix it.
- **Avoid generic advice** unless directly tied to a specific problem visible in the diff.

## What NOT to review
- Files excluded by the "File Selection for gitDiff" rules (do not diff or comment on them).
- Existing unchanged code.
- Overall project structure/architecture unless directly impacted by the changes.
- Missing features or functionality not part of this diff.

## Output Format
Do not include praise or positive feedback.
Only include reviews for actual issues found in the changed code.

Return your review as a JSON object in the following format:
\`\`\`json
{
  "overview": "Summary of specific issues found in the diff changes, 'No issues found', or 'No reviewable changes' if all modified files were excluded.",
  "specificReviews": [
    {
      "file": "path/filename.ext",
      "lines": "N or N-M",
      "review": "Specific issue with the changed code and exact actionable fix."
    }
  ]
}
\`\`\`
`

export type ReviewToolInput = {
  pullRequestTitle?: string
  pullRequestDescription?: string
  commitMessages?: string
  commitRange?: string
  staged?: boolean
  changedFiles?: { path: string; status: string }[]
}

export function formatReviewToolInput(params: ReviewToolInput): string {
  const parts = []
  if (params.pullRequestTitle) {
    parts.push(`<pr_title>\n${params.pullRequestTitle}\n</pr_title>`)
  }
  if (params.pullRequestDescription) {
    parts.push(`<pr_description>\n${params.pullRequestDescription}\n</pr_description>`)
  }
  if (params.commitMessages) {
    parts.push(`<commit_messages>\n${params.commitMessages}\n</commit_messages>`)
  }

  if (params.changedFiles && params.changedFiles.length > 0) {
    const fileList = params.changedFiles.map((file) => `${file.status}: ${file.path}`).join('\n')
    parts.push(`<file_status>\n${fileList}\n</file_status>`)
  }

  let instructions = ''
  if (params.commitRange) {
    instructions = `Review the pull request. Use the gitDiff tool with commit range '${params.commitRange}', contextLines: 5, and includeLineNumbers: true to inspect the actual code changes. The diff will include line number annotations to help you report accurate line numbers. File status information is already provided above.`
  } else if (params.staged) {
    instructions =
      'Review the staged changes. Use the gitDiff tool with staged: true, contextLines: 5, and includeLineNumbers: true to inspect the actual code changes. The diff will include line number annotations to help you report accurate line numbers. File status information is already provided above.'
  } else {
    instructions =
      'Review the unstaged changes. Use the gitDiff tool with contextLines: 5, and includeLineNumbers: true to inspect the actual code changes. The diff will include line number annotations to help you report accurate line numbers. File status information is already provided above.'
  }
  parts.push(`<review_instructions>\n${instructions}\n</review_instructions>`)

  return parts.join('\n')
}

export const ARCHITECT_SYSTEM_PROMPT = `Your task is to create a detailed, step-by-step plan to address the user's request.

## Planning Guidelines

### 1. Understand the Goal
- Thoroughly analyze the user's request to understand their primary objective.
- Use file-system tools (\`listFiles\`, \`readFile\`, \`searchFiles\`) to explore the existing codebase, understand its structure, identify patterns, and gather context. This is crucial for creating a relevant and effective plan.

### 2. Create a Detailed Plan
- Break down the solution into a clear, logical sequence of steps.
- The plan should be a list of concrete, actionable items for a developer to follow.
- Be specific. Instead of "implement the feature," describe the files to be created or modified, the functions or classes to be added, and the logic to be implemented.
- For UI changes, describe the components to be altered and the nature of the changes.

### 3. Plan Structure
- Use a numbered list or a checklist (markdown) to structure the plan.
- Use indentation to show sub-tasks.
- Your final output should be just the plan, without any conversational text.

### 4. Tool Usage
- **listFiles, readFile, searchFiles**: To understand the codebase.
- If the request is ambiguous or lacks necessary details, ask for clarification.

## Your Task

1.  **Analyze the user's request.**
2.  **Explore the codebase** to gather context.
3.  **Create a comprehensive, step-by-step implementation plan.**
4.  **Ask for clarification** if needed.

Please create the plan now.
`

export const CODER_SYSTEM_PROMPT = `${getWorkflowContextPrompt()}**YOU ARE IN PHASE 2: IMPLEMENTATION**

Your task is to implement the plan created and approved in Phase 1.

## Implementation Guidelines

### 1. Plan Analysis

Before starting implementation:
- Review the plan carefully and understand all requirements
- Identify dependencies between different parts of the plan
- Determine if this is a single cohesive task or multiple independent tasks
- Consider the scope and complexity of the work

### 2. Gather Context

Before making changes:
- **Search for similar existing files** to understand patterns and conventions
- **Read relevant files** to see how similar features are implemented
- Look for existing tests, utilities, or helpers you can leverage
- Understand the project structure and naming conventions
- Verify you have all necessary context to proceed

### 3. Implementation Best Practices

- **Make incremental changes**: Implement one piece at a time
- **Follow existing patterns**: Match the style and structure of similar code
- **Add documentation**: Include comments explaining complex logic
- **Consider edge cases**: Think about error handling and boundary conditions
- **Verify as you go**: Test your changes incrementally if possible

### 4. Code Quality

- Follow the project's existing code style and conventions
- Use appropriate TypeScript types (avoid 'any' unless necessary)
- Add JSDoc comments for public APIs and complex functions
- Ensure proper error handling and validation
- Keep functions focused and maintainable

## Your Task

Implement the plan above following these guidelines. Start by:
1. Analyzing the plan structure
2. Searching for similar existing code patterns
3. Proceeding with implementation

Please implement all the necessary code changes according to this plan.
`

export function getImplementPrompt(plan: string): string {
  return `## Your Plan

<plan>
${plan}
</plan>
`
}

export const FIX_SYSTEM_PROMPT = `You are an expert software developer. Your task is to fix a project that is failing a command. You have been provided with the failing command, its output (stdout and stderr), and the exit code. Your goal is to use the available tools to modify the files in the project to make the command pass. Analyze the error, inspect the relevant files, and apply the necessary code changes.

After making changes, you MUST return a JSON object in a markdown block with a summary of the changes you made.
The JSON object must contain a "summary" field, which is a string describing the changes made during the fix attempt.

Example:
\`\`\`json
{
  "summary": "Fixed the 'add' function in 'math.ts' to correctly handle negative numbers."
}
\`\`\`
`

export function getFixUserPrompt(command: string, exitCode: number, stdout: string, stderr: string, task?: string): string {
  const taskContext = task ? `\n## Original Task\n\n${task}\n` : ''

  return `## Context${taskContext}

The following command failed with exit code ${exitCode}:
\`${command}\`

<stdout>
${stdout || '(empty)'}
</stdout>

<stderr>
${stderr || '(empty)'}
</stderr>
`
}
