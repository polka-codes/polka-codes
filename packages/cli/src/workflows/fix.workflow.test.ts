// generated by polka.codes

import { afterEach, describe, expect, mock, spyOn, test } from 'bun:test'
import * as cliShared from '@polka-codes/cli-shared'
import type { WorkflowContext } from '@polka-codes/workflow'
import type { CliToolRegistry } from '../workflow-tools'
import { fixWorkflow } from './fix.workflow'
import * as prompts from './prompts'

const createMockContext = () => {
  const tools = {
    executeCommand: mock<any>(),
    input: mock<any>(),
    generateText: mock<any>(),
    taskEvent: mock<any>(),
    getMemoryContext: mock<any>(async () => ''),
    updateMemory: mock<any>(),
    // Add other tool mocks if needed by the workflow
  }
  const step = mock(async (_name: string, fn: () => any) => fn())
  const logger = {
    info: mock(() => {}),
    error: mock(() => {}),
    warn: mock(() => {}),
    debug: mock(() => {}),
  }

  const context = {
    tools,
    step,
    logger,
  } as unknown as WorkflowContext<CliToolRegistry>

  return { context, tools, step, logger }
}

describe('fixWorkflow', () => {
  afterEach(() => {
    mock.restore()
  })

  test('should succeed when command passes on first attempt', async () => {
    const { context, tools } = createMockContext()
    tools.executeCommand.mockResolvedValue({ exitCode: 0, stdout: 'All tests passed', stderr: '' })

    const result = await fixWorkflow({ command: 'bun test' }, context)

    expect(tools.executeCommand).toHaveBeenCalledWith({ command: 'bun test', shell: true, pipe: true })
    expect(result).toStrictEqual({ success: true, summaries: [] })
  })

  test('should prompt for command when not provided', async () => {
    const { context, tools } = createMockContext()
    const loadConfigSpy = spyOn(cliShared, 'loadConfig').mockReturnValue({
      scripts: {
        check: 'bun typecheck',
        test: 'bun test',
      },
    })

    tools.input.mockResolvedValue('bun typecheck && bun test')
    tools.executeCommand.mockResolvedValue({ exitCode: 0, stdout: 'Success', stderr: '' })

    await fixWorkflow({ interactive: true }, context)

    expect(tools.input).toHaveBeenCalledWith({
      message: 'Please enter the command to run to identify issues:',
      default: 'bun typecheck && bun test',
    })
    expect(tools.executeCommand).toHaveBeenCalledWith({
      command: 'bun typecheck && bun test',
      shell: true,
      pipe: true,
    })

    loadConfigSpy.mockRestore()
  })

  test('should use default command when not interactive', async () => {
    const { context, tools } = createMockContext()
    const loadConfigSpy = spyOn(cliShared, 'loadConfig').mockReturnValue({
      scripts: {
        check: 'bun typecheck',
      },
    })

    tools.executeCommand.mockResolvedValue({ exitCode: 0, stdout: '', stderr: '' })

    await fixWorkflow({ interactive: false }, context)

    expect(tools.executeCommand).toHaveBeenCalledWith({ command: 'bun typecheck', shell: true, pipe: true })

    loadConfigSpy.mockRestore()
  })

  test('should throw error when no command provided and user provides empty input', async () => {
    const { context, tools } = createMockContext()
    const loadConfigSpy = spyOn(cliShared, 'loadConfig').mockReturnValue({})
    tools.input.mockResolvedValue('')

    const promise = fixWorkflow({ interactive: true }, context)

    await expect(promise).rejects.toThrow('No command provided. Aborting.')
    loadConfigSpy.mockRestore()
  })

  test('should succeed after agent fixes the issue', async () => {
    const { context, tools } = createMockContext()

    tools.executeCommand
      .mockResolvedValueOnce({ exitCode: 1, stdout: 'FAIL', stderr: 'Error' })
      .mockResolvedValueOnce({ exitCode: 0, stdout: 'PASS', stderr: '' })

    tools.generateText.mockResolvedValue([
      {
        role: 'assistant',
        content: '```json\n{"summary":"I did a fix","bailReason":null}\n```',
      },
    ])

    const result = await fixWorkflow({ command: 'bun test' }, context)

    expect(tools.executeCommand).toHaveBeenCalledTimes(2)
    expect(tools.generateText).toHaveBeenCalledTimes(1)
    expect(result).toStrictEqual({ success: true, summaries: ['I did a fix'] })
  })

  test('should fail after exhausting all retries', async () => {
    const { context, tools } = createMockContext()

    tools.executeCommand.mockResolvedValue({ exitCode: 1, stdout: 'FAIL', stderr: 'Error' })
    tools.generateText.mockResolvedValue([
      {
        role: 'assistant',
        content: '```json\n{"summary":"I did a fix","bailReason":null}\n```',
      },
    ])

    const result = await fixWorkflow({ command: 'bun test' }, context)

    expect(tools.executeCommand).toHaveBeenCalledTimes(10)
    expect(tools.generateText).toHaveBeenCalledTimes(10)
    expect(result).toStrictEqual({
      success: false,
      reason: 'Failed to fix the issue after maximum attempts.',
      summaries: [
        'I did a fix',
        'I did a fix',
        'I did a fix',
        'I did a fix',
        'I did a fix',
        'I did a fix',
        'I did a fix',
        'I did a fix',
        'I did a fix',
        'I did a fix',
      ],
    })
  })

  test('should return bailReason when agent cannot fix', async () => {
    const { context, tools } = createMockContext()

    tools.executeCommand.mockResolvedValue({ exitCode: 1, stdout: 'FAIL', stderr: 'Mysterious error' })
    tools.generateText.mockResolvedValue([
      {
        role: 'assistant',
        content: '```json\n{"summary":null,"bailReason":"Unable to identify the root cause of the error"}\n```',
      },
    ])

    const result = await fixWorkflow({ command: 'bun test' }, context)

    expect(tools.executeCommand).toHaveBeenCalledTimes(1)
    expect(tools.generateText).toHaveBeenCalledTimes(1)
    expect(result).toStrictEqual({
      success: false,
      summaries: [],
      reason: 'Unable to identify the root cause of the error',
    })
  })

  test('should pass task to agent prompt', async () => {
    const { context, tools } = createMockContext()
    const getFixUserPromptSpy = spyOn(prompts, 'getFixUserPrompt')

    tools.executeCommand.mockResolvedValue({
      exitCode: 1,
      stdout: 'FAIL src/test.ts',
      stderr: 'TypeError: undefined is not a function',
    })
    tools.generateText.mockResolvedValue([
      {
        role: 'assistant',
        content: '```json\n{"summary":"I did a fix"}\n```',
      },
    ])

    tools.generateText.mockResolvedValue([
      {
        role: 'assistant',
        content: '```json\n{"summary":"I did a fix","bailReason":null}\n```',
      },
    ])

    tools.executeCommand
      .mockResolvedValueOnce({ exitCode: 1, stdout: 'FAIL', stderr: 'Error' })
      .mockResolvedValueOnce({ exitCode: 0, stdout: 'PASS', stderr: '' })

    await fixWorkflow({ command: 'bun test', task: 'My original task was to do this thing.' }, context)

    expect(getFixUserPromptSpy).toHaveBeenCalledTimes(1)
    expect(tools.generateText).toHaveBeenCalledTimes(1)

    const generateTextCall = tools.generateText.mock.calls[0][0] as { messages: { role: string; content: string }[] }
    const userMessage = generateTextCall.messages.find((m: { role: string }) => m.role === 'user')
    expect(userMessage?.content).toContain('My original task was to do this thing.')

    getFixUserPromptSpy.mockRestore()
  })
})
