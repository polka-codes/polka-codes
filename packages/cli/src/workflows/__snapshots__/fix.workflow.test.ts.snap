// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`should succeed when command passes on first attempt 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "tool": {
    "input": {
      "command": "bun test",
      "shell": true,
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should succeed when command passes on first attempt 2`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should prompt for command when not provided 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "tool": {
    "input": {
      "default": "bun typecheck && bun test -u",
      "message": "Please enter the command to run to identify issues:",
    },
    "tool": "input",
    "type": "tool",
  },
}
`;

exports[`should prompt for command when not provided 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "command": "bun typecheck && bun test",
      "shell": true,
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should prompt for command when not provided 3`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should use check command as default when only check available 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "tool": {
    "input": {
      "default": "bun typecheck && bun test -u",
      "message": "Please enter the command to run to identify issues:",
    },
    "tool": "input",
    "type": "tool",
  },
}
`;

exports[`should use check command as default when only check available 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "command": "bun typecheck",
      "shell": true,
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should use check command as default when only check available 3`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should use test command as default when only test available 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "tool": {
    "input": {
      "default": "bun typecheck && bun test -u",
      "message": "Please enter the command to run to identify issues:",
    },
    "tool": "input",
    "type": "tool",
  },
}
`;

exports[`should use test command as default when only test available 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "command": "bun test",
      "shell": true,
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should use test command as default when only test available 3`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should combine check and test commands when both available 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "tool": {
    "input": {
      "default": "bun typecheck && bun test -u",
      "message": "Please enter the command to run to identify issues:",
    },
    "tool": "input",
    "type": "tool",
  },
}
`;

exports[`should combine check and test commands when both available 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "command": "eslint . && jest",
      "shell": true,
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should combine check and test commands when both available 3`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should handle object-style script config 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "tool": {
    "input": {
      "default": "bun typecheck && bun test -u",
      "message": "Please enter the command to run to identify issues:",
    },
    "tool": "input",
    "type": "tool",
  },
}
`;

exports[`should handle object-style script config 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "command": "tsc --noEmit && vitest",
      "shell": true,
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should handle object-style script config 3`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should handle no default command 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "tool": {
    "input": {
      "default": "bun typecheck && bun test -u",
      "message": "Please enter the command to run to identify issues:",
    },
    "tool": "input",
    "type": "tool",
  },
}
`;

exports[`should handle no default command 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "command": "npm run build",
      "shell": true,
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should handle no default command 3`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should retry and invoke coder agent when command fails 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "tool": {
    "input": {
      "command": "bun test",
      "shell": true,
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should retry and invoke coder agent when command fails 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "agent": "coder",
      "defaultContext": true,
      "messages": [
        {
          "content": 
"The command "bun test" failed with exit code 1.
Please fix the code.

stdout:
FAIL src/test.ts

stderr:
TypeError: undefined is not a function
"
,
          "type": "user",
        },
      ],
    },
    "tool": "invokeAgent",
    "type": "tool",
  },
}
`;

exports[`should retry and invoke coder agent when command fails 3`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "command": "bun test",
      "shell": true,
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should retry and invoke coder agent when command fails 4`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should retry multiple times before succeeding 1`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should handle command with stdout and stderr 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "agent": "coder",
      "defaultContext": true,
      "messages": [
        {
          "content": 
"The command "make build" failed with exit code 2.
Please fix the code.

stdout:
Building...
Compiling source files

stderr:
Warning: deprecated API
Error: compilation failed
"
,
          "type": "user",
        },
      ],
    },
    "tool": "invokeAgent",
    "type": "tool",
  },
}
`;

exports[`should handle command with stdout and stderr 2`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should throw error after 10 failed attempts 1`] = `
{
  "error": "Failed to fix the issue after 10 attempts.",
  "status": "failed",
}
`;

exports[`should throw error when no command provided and user provides empty input 1`] = `
{
  "error": "No command provided. Aborting.",
  "status": "failed",
}
`;

exports[`should handle command with only stdout 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "agent": "coder",
      "defaultContext": true,
      "messages": [
        {
          "content": 
"The command "echo test" failed with exit code 1.
Please fix the code.

stdout:
Some output

stderr:

"
,
          "type": "user",
        },
      ],
    },
    "tool": "invokeAgent",
    "type": "tool",
  },
}
`;

exports[`should handle command with only stdout 2`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should handle command with only stderr 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "agent": "coder",
      "defaultContext": true,
      "messages": [
        {
          "content": 
"The command "test-cmd" failed with exit code 1.
Please fix the code.

stdout:


stderr:
Error occurred
"
,
          "type": "user",
        },
      ],
    },
    "tool": "invokeAgent",
    "type": "tool",
  },
}
`;

exports[`should handle command with only stderr 2`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should handle config with undefined scripts 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "tool": {
    "input": {
      "default": "bun typecheck && bun test -u",
      "message": "Please enter the command to run to identify issues:",
    },
    "tool": "input",
    "type": "tool",
  },
}
`;

exports[`should handle config with undefined scripts 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "command": "custom-command",
      "shell": true,
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should handle config with undefined scripts 3`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;

exports[`should handle mixed script formats 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "tool": {
    "input": {
      "default": "bun typecheck && bun test -u",
      "message": "Please enter the command to run to identify issues:",
    },
    "tool": "input",
    "type": "tool",
  },
}
`;

exports[`should handle mixed script formats 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "stepFn": [Function],
  "tool": {
    "input": {
      "command": "eslint . && jest --coverage",
      "shell": true,
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should handle mixed script formats 3`] = `
{
  "output": {
    "success": true,
  },
  "status": "completed",
}
`;
